tests/functional/directives.ts

```diff
@@ -0,0 +1,655 @@
+// tslint:disable:member-ordering
+import "reflect-metadata";
+
+import {
+  GraphQLSchema,
+  FieldDefinitionNode,
+  GraphQLField,
+  GraphQLDirective,
+  DirectiveLocation,
+  GraphQLString,
+  parseValue,
+  graphql,
+  GraphQLInputObjectType,
+  InputValueDefinitionNode,
+  InputObjectTypeDefinitionNode,
+  GraphQLInputField,
+  GraphQLScalarType,
+  GraphQLNonNull,
+} from "graphql";
+import {
+  Field,
+  InputType,
+  Resolver,
+  Query,
+  Arg,
+  Directive,
+  buildSchema,
+  ObjectType,
+  Mutation,
+} from "../../src";
+import { getMetadataStorage } from "../../src/metadata/getMetadataStorage";
+import Maybe from "graphql/tsutils/Maybe";
+import { SchemaDirectiveVisitor } from "graphql-tools";
+import { ApolloServer } from "apollo-server";
+
+class UpperCaseType extends GraphQLScalarType {
+  constructor(type: any) {
+    super({
+      name: "UpperCase",
+      parseValue: value => type.parseValue(value),
+      serialize: value => type.serialize(value),
+      parseLiteral: ast => {
+        const result = type.parseLiteral(ast);
+
+        if (typeof result === "string") {
+          return result.toUpperCase();
+        }
+
+        return result;
+      },
+    });
+  }
+}
+
+class UpperCaseDirective extends SchemaDirectiveVisitor {
```

Comment:
Do we really need a custom directive in tests? Can't we use something from npm that works the same?
https://www.npmjs.com/package/@saeris/graphql-directives

Reply:
Yeah, I initially was using 3rd party directives but removed to keep it simple, but I can refactor tests to use 3rd party.

Reply:
We should use 3rd party if exists and if not, just move the them to the helpers file to not make this test suite 1000+ lines big.

Reply:
Also, I did what the docs do for creating directives: https://www.apollographql.com/docs/graphql-tools/schema-directives/

And I was going to use that library you noted but they don't support input types.  :-\

Reply:
Ok, so can we just move it to a separate file?

Reply:
I already did in the latest push.